vortex-array/src/compute/arrays/logical.rs

```diff
@@ -0,0 +1,253 @@
+// SPDX-License-Identifier: Apache-2.0
+// SPDX-FileCopyrightText: Copyright the Vortex contributors
+
+use std::hash::{Hash, Hasher};
+use std::sync::LazyLock;
+
+use enum_map::{Enum, EnumMap, enum_map};
+use futures::{FutureExt, try_join};
+use vortex_buffer::ByteBuffer;
+use vortex_compute::logical::{
+    LogicalAnd, LogicalAndKleene, LogicalAndNot, LogicalOr, LogicalOrKleene,
+};
+use vortex_dtype::DType;
+use vortex_error::VortexResult;
+use vortex_scalar::Scalar;
+use vortex_vector::BoolVector;
+
+use crate::execution::{BatchKernel, BindCtx};
+use crate::serde::ArrayChildren;
+use crate::stats::{ArrayStats, StatsSetRef};
+use crate::vtable::{
+    ArrayVTable, NotSupported, OperatorVTable, SerdeVTable, VTable, VisitorVTable,
+};
+use crate::{
+    Array, ArrayBufferVisitor, ArrayChildVisitor, ArrayEq, ArrayHash, ArrayRef,
+    DeserializeMetadata, EmptyMetadata, EncodingId, EncodingRef, Precision, vtable,
+};
+
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash, Enum)]
+pub enum LogicalOperator {
+    And,
+    AndKleene,
+    Or,
+    OrKleene,
+    AndNot,
+}
+
+vtable!(Logical);
+
+#[derive(Debug, Clone)]
+pub struct LogicalArray {
+    encoding: EncodingRef,
+    lhs: ArrayRef,
+    rhs: ArrayRef,
+    stats: ArrayStats,
+}
+
+impl LogicalArray {
+    /// Create a new logical array.
+    pub fn new(lhs: ArrayRef, rhs: ArrayRef, operator: LogicalOperator) -> Self {
+        assert_eq!(
+            lhs.len(),
+            rhs.len(),
+            "Logical arrays require lhs and rhs to have the same length"
+        );
+
+        // TODO(ngates): should we automatically cast non-null to nullable if required?
+        assert!(matches!(lhs.dtype(), DType::Bool(_)));
+        assert_eq!(lhs.dtype(), rhs.dtype());
+
+        Self {
+            encoding: ENCODINGS[operator].clone(),
+            lhs,
+            rhs,
+            stats: ArrayStats::default(),
+        }
+    }
+
+    /// Returns the operator of this logical array.
+    pub fn operator(&self) -> LogicalOperator {
+        self.encoding.as_::<LogicalVTable>().operator
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct LogicalEncoding {
+    // We include the operator in the encoding so each operator is a different encoding ID.
+    // This makes it easier for plugins to construct expressions and perform pushdown
+    // optimizations.
+    operator: LogicalOperator,
+}
+
+#[allow(clippy::mem_forget)]
```

Comment:
?

Reply:
Yeah, it's a static thing anyway, and it calls mem::forget inside the macro annoyingly.